### PR TITLE
Documentation: Add Packer-ci-build project and Ginkgo Helpers

### PR DESCRIPTION
**Pull request by eloycoto**
Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

### CODE REVIEW

File: Documentation/contributing.rst [hunk 1]
```diff
@@ -60,9 +60,9 @@ To run Cilium locally on VMs, you need:
 +----------------------------------------------------------------------------------+-----------------------+--------------------------------------------------------------------------------+
 | Dependency                                                                       | Version / Commit ID   | Download Command                                                               |
 +==================================================================================+=======================+================================================================================+
-| `Vagrant <https://www.vagrantup.com/downloads.html>`_                            | >= 1.8.3              | `Vagrant Install Instructions <https://www.vagrantup.com/docs/installation/>`_ |
+| `Vagrant <https://www.vagrantup.com/downloads.html>`_                            | >= 2.0                | `Vagrant Install Instructions <https://www.vagrantup.com/docs/installation/>`_ |
```
Comment (joestringer) [inline Documentation/contributing.rst:63]: Doesn't it need to be at least 2.0.2?

File: Documentation/contributing.rst [hunk 2]
```diff
@@ -181,6 +181,23 @@ If for some reason, running of the provisioning script fails, you should bring t
 
     $ vagrant halt
 
+Packer-CI-Build
+^^^^^^^^^^^^^^^^
+
+In cilium, we use a custom base box with a bunch of pre-installed libraries and
```
Comment (ianvernon) [inline Documentation/contributing.rst]: As part of Cilium development, we...

File: Documentation/contributing.rst [hunk 3]
```diff
@@ -181,6 +181,23 @@ If for some reason, running of the provisioning script fails, you should bring t
...
+tools that we need to our daily workflow. That base box is built with `Packer
```
Comment (ianvernon) [inline Documentation/contributing.rst]: ... need to enhance ...

File: Documentation/contributing.rst [hunk 4]
```diff
@@ -181,6 +181,23 @@ If for some reason, running of the provisioning script fails, you should bring t
...
+<https://www.packer.io/>`_ and it's hosted in the `packer-ci-build
```
Comment (ianvernon) [inline Documentation/contributing.rst]: ... It is hosted in the `packer-ci-build <https://github.com/cilium/packer-ci-build>`_ GitHub repository.

File: Documentation/contributing.rst [hunk 5]
```diff
@@ -181,6 +181,23 @@ If for some reason, running of the provisioning script fails, you should bring t
...
+<https://github.com/cilium/packer-ci-build>`_ project.
+
+This base box can be updated and publish from `Jenkins
```
Comment (ianvernon) [inline Documentation/contributing.rst]: New versions of this box can be created via `Jenkins...

File: Documentation/contributing.rst [hunk 6]
```diff
@@ -181,6 +181,23 @@ If for some reason, running of the provisioning script fails, you should bring t
...
+<https://jenkins.cilium.io/job/cilium/job/packer-ci-build/job/master/>`_, where
+a new rebuild will update the `Vagrantcloud <https://app.vagrantup.com/cilium>`_
```
Comment (ianvernon) [inline Documentation/contributing.rst]: ... New builds of the image will be pushed to `Vagrant Cloud...`. The version of the image corresponds to the `BUILD_ID...` on Jenkins.

File: Documentation/contributing.rst [hunk 7]
```diff
@@ -181,6 +181,23 @@ If for some reason, running of the provisioning script fails, you should bring t
...
+and the `BUILD_ID <https://qa.nuxeo.org/jenkins/pipeline-syntax/globals#env>`_
+will be the version.  That version ID will be used in Cilium `Vagrantfiles
+<https://github.com/cilium/cilium/blob/master/test/Vagrantfile#L10>`_.
+
+In case that we want to make changes in packer-ci-build project, each PR in
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Change to the following:

Changes to this image are made via contributions to the `packer-ci-build` repository. Authorized GitHub users can trigger builds with a GitHub comment on the PR containing the trigger phrase ``build-me-please`` .

File: Documentation/contributing.rst [hunk 8]
```diff
@@ -565,6 +582,27 @@ Best Practices for Writing Tests
         vm.ReportFailed()
     }
 
+
+Ginkgo Extensions
+~~~~~~~~~~~~~~~~~~
+
+In cilium, some Ginkgo features are extended to cover some uses cases that are
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Make sure to capitalize Cilium.

File: Documentation/contributing.rst [hunk 9]
```diff
@@ -565,6 +582,27 @@ Best Practices for Writing Tests
...
+useful for how cilium run test
```
Comment (ianvernon) [inline Documentation/contributing.rst]: change to "useful for testing Cilium."

File: Documentation/contributing.rst [hunk 10]
```diff
@@ -565,6 +582,27 @@ Best Practices for Writing Tests
...
+
+BeforeAll
+^^^^^^^^^^
+
+This function will run before all BeforeEach within a describe or context. This
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Link to the corresponding Ginkgo constructs (BeforeEach, Describe, Context).

File: Documentation/contributing.rst [hunk 11]
```diff
@@ -565,6 +582,27 @@ Best Practices for Writing Tests
...
+function is the new ``SetUp`` or initalize function for our Ginkgo test.
```
Comment (ianvernon) [inline Documentation/contributing.rst]: What is ``SetUp``? Is this a Ginkgo construct?

File: Documentation/contributing.rst [hunk 12]
```diff
@@ -565,6 +582,27 @@ Best Practices for Writing Tests
...
+
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Would be useful to provide an example in the Ginkgo extensions section to show what these constructs do in relation to each other (i.e., a dummy test with multiple `It`s, a `BeforeAll`, `AfterAll`, `AfterEach`, `BeforeEach`, etc. to show users in what order the various things run, like;


```Describe --> BeforeAll --> BeforeEach --> It --> AfterEach --> BeforeEach --> It --> AfterEach --> AfterAll```

File: Documentation/contributing.rst [hunk 13]
```diff
@@ -565,6 +582,27 @@ Best Practices for Writing Tests
...
+AfterAll
+^^^^^^^^^
+
+This function will run after all ``AfterEeach`` functions defined in a Describe or
```
Comment (ianvernon) [inline Documentation/contributing.rst]: extra E in `AfterEach`. Link to `AfterEach` Ginkgo documentation, as well as Describe, Context, etc.

File: Documentation/contributing.rst [hunk 14]
```diff
@@ -565,6 +582,27 @@ Best Practices for Writing Tests
...
+Context. This function is the ``tearDown`` or the ``finilize`` function. The best
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Using \`\`\ makes it seem like the contents between the ticks are official terms, but they aren't. Remove ``tearDown`` and ``finilize`` and say something like:

This is used for tearing down objects created which are used by all `It`s within the given `Context` or `Describe`.  It is ran after all `It`s have ran.

Comment (ianvernon) [inline Documentation/contributing.rst]: Change to: A good use case for using `AfterAll` is to remove containers or pods that are needed for multiple `It`s.

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: Documentation/contributing.rst (hunk 14 above)
Comment (joestringer) [inline Documentation/contributing.rst]: `finilize` -> `finalize`

File: Documentation/contributing.rst [hunk 15]
```diff
@@ -181,6 +181,23 @@ If for some reason, running of the provisioning script fails, you should bring t
 
     $ vagrant halt
 
+Packer-CI-Build
```
Comment (joestringer) [inline Documentation/contributing.rst]: As a developer, when do I care about Packer-CI-Build? How do I know if I should trigger a new build, and how do I trigger it?

File: Documentation/contributing.rst (hunk 1 above)
Comment (eloycoto) [reply Documentation/contributing.rst:63]: 2.0 is ok ;-)

File: Documentation/contributing.rst (hunk 15 above)
Comment (eloycoto) [reply Documentation/contributing.rst]: Done

**Review by manalibhutiyani — APPROVED** [review]

File: Documentation/contributing.rst [hunk 16]
```diff
@@ -181,6 +181,31 @@ If for some reason, running of the provisioning script fails, you should bring t
...
+the `BUILD_ID <https://qa.nuxeo.org/jenkins/pipeline-syntax/globals#env>`_ on
+Jenkins job.  That version ID will be used in Cilium `Vagrantfiles
+<https://github.com/cilium/cilium/blob/master/test/Vagrantfile#L10>`_.
+
+Changes to this image are made via contributions to the packer-ci-build
+repository. Authorized GitHub users can trigger builds with a GitHub comment on
+the PR containing the trigger phrase ``build-me-please``.
+
+This box will need to be updated when a new developer needs a new dependency
+that it's not satisfied in the current version of the box, or the cilium
```
Comment (ianvernon) [inline Documentation/contributing.rst]: ... that is not installed in the current version of the box, or if a dependency that is cached within the box becomes stale.

File: Documentation/contributing.rst [hunk 17]
```diff
@@ -181,6 +181,31 @@ If for some reason, running of the provisioning script fails, you should bring t
...
+pre-installed libraries and tools that we need to enhance our daily workflow.
+That base box is built with `Packer <https://www.packer.io/>`_ and it is hosted
+in the `packer-ci-build <https://github.com/cilium/packer-ci-build>`_ Github
+repository.
+
+New versions of this box can be created via `Jenkins
+<https://jenkins.cilium.io/job/cilium/job/packer-ci-build/job/master/>`_, where
+new builds of the image will be pushed to  `Vagrant Cloud
+<https://app.vagrantup.com/cilium>`_ . The version of the image correspods to
+the `BUILD_ID <https://qa.nuxeo.org/jenkins/pipeline-syntax/globals#env>`_ on
```
Comment (ianvernon) [inline Documentation/contributing.rst]: ... environemtn variable in the Jenkins job.

File: Documentation/contributing.rst [hunk 18]
```diff
@@ -181,6 +181,31 @@ If for some reason, running of the provisioning script fails, you should bring t
 
     $ vagrant halt
 
+Packer-CI-Build
+^^^^^^^^^^^^^^^^
+
+As part of Cilium development, we use a custom base box with a bunch of
+pre-installed libraries and tools that we need to enhance our daily workflow.
+That base box is built with `Packer <https://www.packer.io/>`_ and it is hosted
+in the `packer-ci-build <https://github.com/cilium/packer-ci-build>`_ Github
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Github --> GitHub

File: Documentation/contributing.rst [hunk 19]
```diff
@@ -565,6 +590,38 @@ Best Practices for Writing Tests
...
+useful for testing Cilium.
+
+BeforeAll
+^^^^^^^^^^
+
+This function will run before all `BeforeEach
+<https://onsi.github.io/ginkgo/#extracting-common-setup-beforeeach>`_ within a
+`Describe or Context
+<https://onsi.github.io/ginkgo/#organizing-specs-with-containers-describe-and-context>`_.
+This method is an equivalent to ``SetUp`` or initalize function in common
```
Comment (ianvernon) [inline Documentation/contributing.rst]: function --> functions

**Review by ianvernon — CHANGES_REQUESTED** [review]

**Comment by eloycoto** [comment]
@ianvernon comments sorted, ready to review again.

**Comment by eloycoto** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Pull request merged**